text/0000-fn-lifetime.md

```diff
@@ -0,0 +1,88 @@
+- Feature Name: fn lifetime
+- Start Date: 2017-01-02
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a `'fn` lifetime that is bound to the scope of the body of the current innermost function or closure.
+
+# Motivation
+[motivation]: #motivation
+
+Doing this will enable us to declare that some values live longer than their containing scope, allowing us to "allocate" recursive data structures (e.g. linked lists) on the stack, as in:
+
+```rust
+struct Node<'a, T> {
+    data: T,
+    next: Option<&'a Node>,
+}
+
+fn foo() {
+    // here we ascribe the `'fn` lifetime to `next`
+    let mut head : Node<'fn, usize> = Node { data: 0, None };
+    for i in iter {
+        head = Node { data: i, Some(&head) }
+    }
+}
+```
+
+# Detailed design
+[design]: #detailed-design
+
+There are three steps
+
+1. Bind `'fn` in every closure or function body, representing the lifetime of the body.
+   This is equivalent to adding an additional `'fn` lifetime parameter to every function that cannot be used in the function's type (argument declaration or where clause), nor manually instantiated by a caller.
+   Closure literals currently do not allow explicit parameters, but the desugaring would be the same if they did.
+   Note that the `'fn` lifetimes will shadow each other, whereas we don't currently allow lifetime shadowing.
+
+2. Allow items inside functions to use lifetimes, including `'fn`, bound by the enclosing functions.
+   Unclear how this interfacts with current well-formedness rules.
+
+3. Change axioms of borrowing so it is fine if the lifetime is already entered, as long as the borrow does not outlive it *from* the point of borrowing *after*. For example:
```

Comment:
@nikomatsakis you are correct. I had noticed the mistake but hadn't gotten around to fixing it---sorry to have wasted your time with that.